README.md
KP_PointCache-Reader-ICE-Node
=============================

Source Code for Softimage ICE Node to read a pc2 file.


CHANGELOG.md
Change Log v 1.1:
Bug fixes:

·	Point data for files over 2 GB was not read correctly due to limitations of <int> data type. It is changed to <__int64> which can handle the data > 2 GB. Now the node can easily read files over 2 GB.

Notes
·	The same changes needs to be applied for KP_PointCacheReader for this bugfix. I am not aware if Kai has fixed this already in other updates to his code. But the one I had from rray.de does not have this change.

·	The code for this bugfix is not supported on Linux. However, in case you want to compile for Linux, please note that the <__int64> data type for windows translates to <long long> on a gcc compiler for Linux. I will add support for Linux once I have a machine with Linux up and running. Till then you have to change the code yourself.


·	Suppressed unnecessary warning messages in cases of pc2 file not specified and unable to open file. They were kind of annoying. Just uncomment my code lines to bring them back if you want.


New features:

·	When working with pc2 files I always missed the ability to know beforehand the start and end frame of the file. Now the node supports this. There are two new output ports that furnish the start and end frames as scalars. You can use this to do time warps, offsets, view them in viewports as custom attributes or whatever else you might find it useful for.



HelperFunctions.h
//#########################################################################################
//#########################################################################################
//                                                                                       ##
//  KP_PointCache-Reader-ICE-Node V 1.0 (c) 2013 Alok Gandhi (alok.gandhi2002@gmail.com) ##
//                                                                                       ##
//                                                                                       ##
//  This file is part of KP_PointCache-Reader-ICE-Node.                                  ##
//                                                                                       ##
//  KP_PointCache-Reader-ICE-Node is free software: you can redistribute it and/or modify##
//  it under the terms of the GNU General Public License, Version 3, 29 June 2007        ##
//  as published by the Free Software Foundation,                                        ##
//                                                                                       ##
//  KP_PointCache-Reader-ICE-Node is distributed in the hope that it will be useful,     ##
//  but WITHOUT ANY WARRANTY; without even the implied warranty of                       ##
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                        ##
//  GNU General Public License for more details.                                         ##
//                                                                                       ##
//  You should have received a copy of the GNU General Public License                    ##
//  along with KP_PointCache-Reader-ICE-Node.  If not, see <http://www.gnu.org/licenses/>##
//	                                                                                     ##
//	The original Code for KP_PointCache Core is written by Kai Wolter. This work is just ##
//	an extension / reuse of his code as is.                                              ##
//	                                                                                     ##
//	Kai Wolter's original Licence can be found in the other files of this code as        ##
//	and when                                                                             ##
//	needed.                                                                              ##
//                                                                                       ##
//                                                                                       ##
//#########################################################################################
//#########################################################################################


#ifndef _HelperFunctions_
#define _HelperFunctions_

#include <xsi_application.h>
#include <xsi_context.h>
#include <xsi_pluginregistrar.h>
#include <xsi_status.h>

#include <xsi_icenodecontext.h>
#include <xsi_icenodedef.h>
#include <xsi_command.h>
#include <xsi_factory.h>
#include <xsi_longarray.h>
#include <xsi_doublearray.h>
#include <xsi_math.h>
#include <xsi_vector2f.h>
#include <xsi_vector3f.h>
#include <xsi_vector4f.h>
#include <xsi_matrix3f.h>
#include <xsi_matrix4f.h>
#include <xsi_rotationf.h>
#include <xsi_quaternionf.h>
#include <xsi_color4f.h>
#include <xsi_shape.h>
#include <xsi_icegeometry.h>
#include <xsi_iceportstate.h>
#include <xsi_indexset.h>
#include <xsi_dataarray.h>
#include <xsi_dataarray2D.h>
#include <xsi_ppglayout.h>

#include "KP_PointCacheCore.h"


using namespace XSI;
using namespace MATH;

double LinearInterpolate(
   double y1,double y2,
   double mu);

double CosineInterpolate(
   double y1,double y2,
   double mu);

double CubicInterpolate(
   double y0,double y1,
   double y2,double y3,
   double mu);

double BezierInterpolate(
   double y0,double y1,
   double y2,double y3,
   double mu);

double HermiteInterpolate(
   double y0,double y1,
   double y2,double y3,
   double mu,
   double tension,
   double bias);

void ReadFrameData(PointCacheCore *pCacheCore, int frameNumber, bool useYUp, int numDataPoints, CVector3Array & returnArray);

void CalculatePosArray(PointCacheCore *pCacheCore, double currentFrame, double timeWarp, double offset, CVector3Array &posArray, bool useYUp, int interpolation, double hermite_tension, double hermite_bias);


#endif

HelperFunctions.cpp
//#########################################################################################
//#########################################################################################
//                                                                                       ##
//  KP_PointCache-Reader-ICE-Node V 1.0 (c) 2013 Alok Gandhi (alok.gandhi2002@gmail.com) ##
//                                                                                       ##
//                                                                                       ##
//  This file is part of KP_PointCache-Reader-ICE-Node.                                  ##
//                                                                                       ##
//  KP_PointCache-Reader-ICE-Node is free software: you can redistribute it and/or modify##
//  it under the terms of the GNU General Public License, Version 3, 29 June 2007        ##
//  as published by the Free Software Foundation,                                        ##
//                                                                                       ##
//  KP_PointCache-Reader-ICE-Node is distributed in the hope that it will be useful,     ##
//  but WITHOUT ANY WARRANTY; without even the implied warranty of                       ##
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                        ##
//  GNU General Public License for more details.                                         ##
//                                                                                       ##
//  You should have received a copy of the GNU General Public License                    ##
//  along with KP_PointCache-Reader-ICE-Node.  If not, see <http://www.gnu.org/licenses/>##
//	                                                                                     ##
//	The original Code for KP_PointCache Core is written by Kai Wolter. This work is just ##
//	an extension / reuse of his code as is.                                              ##
//	                                                                                     ##
//	Kai Wolter's original Licence can be found in the other files of this code as        ##
//	and when                                                                             ##
//	needed.                                                                              ##
//                                                                                       ##
//                                                                                       ##
//#########################################################################################
//#########################################################################################

#include "HelperFunctions.h"



/// linear interpolation implementation
double LinearInterpolate(
   double y1,double y2,
   double mu)
{
   return(y1*(1-mu)+y2*mu);
}

/// cosine interpolation implementation
double CosineInterpolate(
   double y1,double y2,
   double mu)
{
   double mu2;

   mu2 = (1-cos(mu*PI))/2;
   return(y1*(1-mu2)+y2*mu2);
}

/// cubic interpolation implementation
double CubicInterpolate(
   double y0,double y1,
   double y2,double y3,
   double mu)
{
   double a0,a1,a2,a3,mu2;

   mu2 = mu*mu;
   a0 = y3 - y2 - y0 + y1;
   a1 = y0 - y1 - a0;
   a2 = y2 - y0;
   a3 = y1;

   return(a0*mu*mu2+a1*mu2+a2*mu+a3);
}

/// bezier interpolation implementation
double BezierInterpolate(
   double y0,double y1,
   double y2,double y3,
   double mu)
{
   double mum1,mum13,mu3;


   mum1 = 1 - mu;
   mum13 = mum1 * mum1 * mum1;
   mu3 = mu * mu * mu;

   return ( mum13*y0 + 3*mu*mum1*mum1*y1 + 3*mu*mu*mum1*y2 + mu3*y3 );
}

/// hermite interpolation implementation
double HermiteInterpolate(
   double y0,double y1,
   double y2,double y3,
   double mu,
   double tension,
   double bias)
{
   double m0,m1,mu2,mu3;
   double a0,a1,a2,a3;

	mu2 = mu * mu;
	mu3 = mu2 * mu;
   m0  = (y1-y0)*(1+bias)*(1-tension)/2;
   m0 += (y2-y1)*(1-bias)*(1-tension)/2;
   m1  = (y2-y1)*(1+bias)*(1-tension)/2;
   m1 += (y3-y2)*(1-bias)*(1-tension)/2;
   a0 =  2*mu3 - 3*mu2 + 1;
   a1 =    mu3 - 2*mu2 + mu;
   a2 =    mu3 -   mu2;
   a3 = -2*mu3 + 3*mu2;

   return(a0*y1+a1*m0+a2*m1+a3*y2);
}

/// addresses data and reads from file
void ReadFrameData(PointCacheCore *pCacheCore, int frameNumber, bool useYUp, int numDataPoints, CVector3Array & returnArray)
{
	pCacheCore->AddressData(frameNumber, numDataPoints );	
	for (int i=0; i<numDataPoints; ++i )
	{	
			double x,y,z, helper;
			pCacheCore->ReadDouble(x);
			pCacheCore->ReadDouble(y);
			pCacheCore->ReadDouble(z);

			if ( ! useYUp )
			{
				helper = z;
				z = -y;
				y = helper;
			}

			returnArray[i].Set( x , y, z );
	}
}

/// Calculates final vertex positions based on frame, timewarp, offset, influence and interpolations parameters
void CalculatePosArray(PointCacheCore *pCacheCore, double currentFrame, double timeWarp, double offset, CVector3Array &posArray, bool useYUp, int interpolation, double hermite_tension, double hermite_bias)
{
	int numDataPoints = (int) posArray.GetCount();

	//pCacheCore->GetNumPoints(numDataPoints);
	double sampleRate;
	pCacheCore->GetSampleRate(sampleRate);
	int numSamples;
	pCacheCore->GetNumSamples(numSamples);
	double startTime;
	pCacheCore->GetStartTime(startTime);

	currentFrame = ( currentFrame - offset ) * timeWarp;

	if ( currentFrame < startTime )
		currentFrame = startTime;

	double frameInBuffer = (currentFrame - startTime) * (1/sampleRate);

	

	if ( frameInBuffer > numSamples -1 )
	  frameInBuffer = numSamples -1;

	int firstFrameInBuffer = (int) floor(frameInBuffer);
	double nextFrameRatio = frameInBuffer - firstFrameInBuffer;

	if ( nextFrameRatio == 0.0 ) // simplest case no interpolation required
	  ReadFrameData(pCacheCore, firstFrameInBuffer, useYUp, numDataPoints, posArray);
	else
	{
		CVector3Array current_posArray(numDataPoints); // current positions
		CVector3Array next_posArray(numDataPoints); // current positions
		CVector3Array before_posArray(numDataPoints); // before positions
		CVector3Array after_posArray(numDataPoints); // after positions

		int nextFrameInBuffer = (int) ceil(frameInBuffer);
		// we need interpolation and maybe need the point positions before and after as well
		if (interpolation > 1 ) //cubic/hermite/bezier
		{
			int beforeFrameInBuffer = max(0, firstFrameInBuffer - 1);
			int afterFrameInBuffer = min(numSamples -1, nextFrameInBuffer  + 1);

			// fill before array
			ReadFrameData(pCacheCore, beforeFrameInBuffer, useYUp, numDataPoints, before_posArray);
			// fill after array
			ReadFrameData(pCacheCore, afterFrameInBuffer, useYUp, numDataPoints, after_posArray);
		}

		ReadFrameData(pCacheCore, firstFrameInBuffer, useYUp, numDataPoints, current_posArray);
		ReadFrameData(pCacheCore, nextFrameInBuffer, useYUp, numDataPoints, next_posArray);
		
		// all needed data acquired, let's do interpolation
		double x,y,z;
		
		if ( interpolation == 0 ) // linear
		{
			for (int i=0; i<numDataPoints; ++i )
			{	
				x = LinearInterpolate(current_posArray[i].GetX(),next_posArray[i].GetX(), nextFrameRatio);
				y = LinearInterpolate(current_posArray[i].GetY(),next_posArray[i].GetY(), nextFrameRatio);
				z = LinearInterpolate(current_posArray[i].GetZ(),next_posArray[i].GetZ(), nextFrameRatio);
				posArray[i].Set( x , y, z );	
			}
		}
		else
		if ( interpolation == 1 ) // cosine
		{
			for (int i=0; i<numDataPoints; ++i )
			{	
				x = CosineInterpolate(current_posArray[i].GetX(),next_posArray[i].GetX(), nextFrameRatio);
				y = CosineInterpolate(current_posArray[i].GetY(),next_posArray[i].GetY(), nextFrameRatio);
				z = CosineInterpolate(current_posArray[i].GetZ(),next_posArray[i].GetZ(), nextFrameRatio);
				posArray[i].Set( x , y, z );	
			}
		}
		else
		if ( interpolation == 2 ) // cubic
		{
			for (int i=0; i<numDataPoints; ++i )
			{
				x = CubicInterpolate(before_posArray[i].GetX(), current_posArray[i].GetX(),next_posArray[i].GetX(), after_posArray[i].GetX(), nextFrameRatio);
				y = CubicInterpolate(before_posArray[i].GetY(), current_posArray[i].GetY(),next_posArray[i].GetY(), after_posArray[i].GetY(), nextFrameRatio);
				z = CubicInterpolate(before_posArray[i].GetZ(), current_posArray[i].GetZ(),next_posArray[i].GetZ(), after_posArray[i].GetZ(), nextFrameRatio);
				posArray[i].Set( x , y, z );	
			}
		}
		else
		if ( interpolation == 3 ) // hermite
		{
			for (int i=0; i<numDataPoints; ++i )
			{
				x = HermiteInterpolate(before_posArray[i].GetX(), current_posArray[i].GetX(),next_posArray[i].GetX(), after_posArray[i].GetX(), nextFrameRatio, hermite_tension, hermite_bias);
				y = HermiteInterpolate(before_posArray[i].GetY(), current_posArray[i].GetY(),next_posArray[i].GetY(), after_posArray[i].GetY(), nextFrameRatio, hermite_tension, hermite_bias);
				z = HermiteInterpolate(before_posArray[i].GetZ(), current_posArray[i].GetZ(),next_posArray[i].GetZ(), after_posArray[i].GetZ(), nextFrameRatio, hermite_tension, hermite_bias);
				posArray[i].Set( x , y, z );
			}
		}
		else
		if ( interpolation == 4 ) // bezier for future use
		{
			for (int i=0; i<numDataPoints; ++i )
			{
				x = BezierInterpolate(before_posArray[i].GetX(), current_posArray[i].GetX(),next_posArray[i].GetX(), after_posArray[i].GetX(),  nextFrameRatio);
				y = BezierInterpolate(before_posArray[i].GetY(), current_posArray[i].GetY(),next_posArray[i].GetY(), after_posArray[i].GetY(),  nextFrameRatio);
				z = BezierInterpolate(before_posArray[i].GetZ(), current_posArray[i].GetZ(),next_posArray[i].GetZ(), after_posArray[i].GetZ(),  nextFrameRatio);
				posArray[i].Set( x , y, z );
			}
		}
	}
}



KP_PointCacheCore.h
// ===========================================================================
//
//	KP_PointCacheCore.h
//
//	Copyright (C) 2007-2008, Kai Wolter
//	http://www.kai-wolter.com
//	mail@kai-wolter.com
//
//	Created: 2008/01/20 (y/m/d)
//
//	--------------------------------------------------------------------------
//
//	This program is free software: you can redistribute it and/or modify
//	it under the terms of the GNU General Public License as published by
//   	the Free Software Foundation, either version 3 of the License, or
//	(at your option) any later version.
//
//	This program is distributed in the hope that it will be useful,
//    	but WITHOUT ANY WARRANTY; without even the implied warranty of
//    	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    	GNU General Public License for more details.
//
//	You should have received a copy of the GNU General Public License
//	along with this program; if not, see <http://www.gnu.org/licenses/>.
//
// ===========================================================================

#ifndef _PointCacheCore_
#define _PointCacheCore_

#include <fstream>
using namespace std;

#define LOAD_OPT_STREAM 0 
#define LOAD_OPT_FRAME 1
#define LOAD_OPT_BUFFER 2

#define SIZE_FLOAT 4
#define SIZE_INT 4

struct PointCacheHeader
{
	char* signature;	// cache file type
    int version;		// version number
	int numPoints;		// number of stored points (vertices)
	float startTime;	// first recorded frame
	float sampleRate;   // samples per frame
	int numSamples;		// number of recorded samples
};

class PointCacheCore
{
private:
	PointCacheHeader* header;

	ifstream infile;	
	ofstream outfile;	
	int dataSize;
	char * buffer;
	int pcType;
	int bufferAddress;
	char cacheFilename[500];

public:
	PointCacheCore();
	int OpenCache(char* filename, int loadOption = 0);

	int ReadFloat(float& value);
	int ReadInt(int& value);
	int ReadDouble(double& value);

	int WriteCache(char* filename, int numPoints, float startFrame, float sampleRate, int numSamples, int pcType = 0);

	int Write(float data);
	int Write(int data);
	int Write(double data);

	int Close();

	int AddressData(int frame, int points);

	int GetVersion(int& value);
	int GetNumPoints(int& value);
	int GetStartTime(double& value);
	int GetSampleRate(double& value);
	int GetNumSamples(int& value);
	
	void DeleteBuffer();
	
	bool IsReadingCache();
	bool IsWritingCache();
	
	int GetPCType();
};

#endif

plugin.cpp
//#########################################################################################
//#########################################################################################
//                                                                                       ##
//  KP_PointCache-Reader-ICE-Node V 1.0 (c) 2013 Alok Gandhi (alok.gandhi2002@gmail.com) ##
//                                                                                       ##
//                                                                                       ##
//  This file is part of KP_PointCache-Reader-ICE-Node.                                  ##
//                                                                                       ##
//  KP_PointCache-Reader-ICE-Node is free software: you can redistribute it and/or modify##
//  it under the terms of the GNU General Public License, Version 3, 29 June 2007        ##
//  as published by the Free Software Foundation,                                        ##
//                                                                                       ##
//  KP_PointCache-Reader-ICE-Node is distributed in the hope that it will be useful,     ##
//  but WITHOUT ANY WARRANTY; without even the implied warranty of                       ##
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                        ##
//  GNU General Public License for more details.                                         ##
//                                                                                       ##
//  You should have received a copy of the GNU General Public License                    ##
//  along with KP_PointCache-Reader-ICE-Node.  If not, see <http://www.gnu.org/licenses/>##
//	                                                                                     ##
//	The original Code for KP_PointCache Core is written by Kai Wolter. This work is just ##
//	an extension / reuse of his code as is.                                              ##
//	                                                                                     ##
//	Kai Wolter's original Licence can be found in the other files of this code as        ##
//	and when                                                                             ##
//	needed.                                                                              ##
//                                                                                       ##
//                                                                                       ##
//#########################################################################################
//#########################################################################################



#include "HelperFunctions.h"

using namespace XSI;
using namespace MATH;

enum IDs
{
	ID_IN_File_Name = 0,
	ID_IN_Offset = 1,
	ID_IN_TimeWarp = 2,
	ID_IN_UseCustomPlayback = 3,
	ID_IN_Frame = 4,
	ID_IN_Interpolation = 5,
	ID_IN_HermiteTension = 6,
	ID_IN_HermiteBias = 7,
	ID_IN_LoadOption = 8,
	ID_IN_UseYup = 9,
	ID_G_100 = 100,
	ID_OUT_Points = 200,
	ID_OUT_StartFrame = 201,
	ID_OUT_EndFrame = 202,
	ID_TYPE_CNS = 400,
	ID_STRUCT_CNS,
	ID_CTXT_CNS,
	ID_UNDEF = ULONG_MAX
};

XSI::CStatus RegisterReadPC2( XSI::PluginRegistrar& in_reg );



SICALLBACK XSILoadPlugin( PluginRegistrar& in_reg )
{
	in_reg.PutAuthor(L"Alok Gandhi");
	in_reg.PutName(L"ReadPC2");
	in_reg.PutVersion(1,1);

	RegisterReadPC2( in_reg );

	return CStatus::OK;
}

SICALLBACK XSIUnloadPlugin( const PluginRegistrar& in_reg )
{
	CString strPluginName;
	strPluginName = in_reg.GetName();
	Application().LogMessage(strPluginName + L" has been unloaded.",siVerboseMsg);
	return CStatus::OK;
}


CStatus RegisterReadPC2( PluginRegistrar& in_reg )
{
	ICENodeDef nodeDef;
	nodeDef = Application().GetFactory().CreateICENodeDef(L"ReadPC2",L"ReadPC2");

	CStatus st;
	st = nodeDef.PutColor(154,188,102);
	st.AssertSucceeded( ) ;


	st = nodeDef.AddPortGroup(ID_G_100);
	st.AssertSucceeded( ) ;

	st = nodeDef.AddInputPort(	ID_IN_File_Name,
								ID_G_100,
								siICENodeDataString,
								siICENodeStructureSingle,
								siICENodeContextAny,
								L"File_Name",
								L"File_Name",
								L"",
								CValue(),
								CValue(),
								ID_UNDEF,
								ID_UNDEF,
								ID_CTXT_CNS);
	st.AssertSucceeded( ) ;

	
	
	st = nodeDef.AddInputPort(	ID_IN_Offset,ID_G_100,
								siICENodeDataFloat,
								siICENodeStructureSingle,
								siICENodeContextAny,
								L"Offset",
								L"Offset",
								0,
								CValue(),
								CValue(),
								ID_UNDEF,
								ID_UNDEF,
								ID_CTXT_CNS);
	st.AssertSucceeded( ) ;

	
	
	st = nodeDef.AddInputPort(	ID_IN_TimeWarp,
								ID_G_100,siICENodeDataFloat,
								siICENodeStructureSingle,
								siICENodeContextAny,
								L"TimeWarp",
								L"TimeWarp",
								1.0f,
								CValue(),
								CValue(),
								ID_UNDEF,
								ID_UNDEF,
								ID_CTXT_CNS);
	st.AssertSucceeded( ) ;

	
	
	st = nodeDef.AddInputPort(	ID_IN_UseCustomPlayback,
								ID_G_100,
								siICENodeDataBool,
								siICENodeStructureSingle,
								siICENodeContextAny,
								L"UseCustomPlayback",
								L"UseCustomPlayback",
								false,
								CValue(),
								CValue(),
								ID_UNDEF,
								ID_UNDEF,
								ID_CTXT_CNS);
	st.AssertSucceeded( ) ;

	
	
	st = nodeDef.AddInputPort(	ID_IN_Frame,
								ID_G_100,
								siICENodeDataFloat,
								siICENodeStructureSingle,
								siICENodeContextAny,
								L"Frame",
								L"Frame",
								0,
								CValue(),
								CValue(),
								ID_UNDEF,
								ID_UNDEF,
								ID_CTXT_CNS);
	st.AssertSucceeded( ) ;

	
	
	st = nodeDef.AddInputPort(	ID_IN_Interpolation,
								ID_G_100,
								siICENodeDataLong,
								siICENodeStructureSingle,
								siICENodeContextAny,
								L"Interpolation",
								L"Interpolation",
								0,
								CValue(),
								CValue(),
								ID_UNDEF,
								ID_UNDEF,
								ID_CTXT_CNS);
	st.AssertSucceeded( ) ;

	
	
	st = nodeDef.AddInputPort(	ID_IN_HermiteTension,
								ID_G_100,
								siICENodeDataFloat,
								siICENodeStructureSingle,
								siICENodeContextAny,
								L"HermiteTension",
								L"HermiteTension",
								0.0f,
								CValue(),
								CValue(),
								ID_UNDEF,
								ID_UNDEF,
								ID_CTXT_CNS);
	st.AssertSucceeded( ) ;

	
	
	st = nodeDef.AddInputPort(	ID_IN_HermiteBias,
								ID_G_100,
								siICENodeDataFloat,
								siICENodeStructureSingle,
								siICENodeContextAny,
								L"HermiteBias",
								L"HermiteBias",
								0.0f,
								CValue(),
								CValue(),
								ID_UNDEF,
								ID_UNDEF,
								ID_CTXT_CNS);
	st.AssertSucceeded( ) ;

	
	
	st = nodeDef.AddInputPort(	ID_IN_LoadOption,
								ID_G_100,
								siICENodeDataLong,
								siICENodeStructureSingle,
								siICENodeContextAny,
								L"LoadOption",
								L"LoadOption",
								1,
								CValue(),
								CValue(),
								ID_UNDEF,
								ID_UNDEF,
								ID_CTXT_CNS);
	st.AssertSucceeded( ) ;

	
	
	st = nodeDef.AddInputPort(	ID_IN_UseYup,
								ID_G_100,
								siICENodeDataBool,
								siICENodeStructureSingle,
								siICENodeContextAny,
								L"UseYup",
								L"UseYup",
								false,
								CValue(),
								CValue(),
								ID_UNDEF,
								ID_UNDEF,
								ID_CTXT_CNS);
	st.AssertSucceeded( ) ;

	// Output Ports
	st = nodeDef.AddOutputPort(	ID_OUT_Points,
								siICENodeDataVector3,
								siICENodeStructureArray,
								siICENodeContextAny,
								L"Points",
								L"Points",
								ID_UNDEF,
								ID_UNDEF,
								ID_CTXT_CNS);
	st.AssertSucceeded( ) ;

	st = nodeDef.AddOutputPort(	ID_OUT_StartFrame,
								siICENodeDataFloat,
								siICENodeStructureSingle,
								siICENodeContextAny,
								L"StartFrame",
								L"StartFrame",
								ID_UNDEF,
								ID_UNDEF,
								ID_CTXT_CNS);
	st.AssertSucceeded( ) ;

	st = nodeDef.AddOutputPort(	ID_OUT_EndFrame,
								siICENodeDataFloat,
								siICENodeStructureSingle,
								siICENodeContextAny,
								L"EndFrame",
								L"EndFrame",
								ID_UNDEF,
								ID_UNDEF,
								ID_CTXT_CNS);
	st.AssertSucceeded( ) ;

	
	
	PluginItem nodeItem = in_reg.RegisterICENode(nodeDef);
	nodeItem.PutCategories(L"Custom ICENode");

	return CStatus::OK;
}

SICALLBACK ReadPC2_Init( CRef& in_ref )
{
	Context ctxt = in_ref;

	PointCacheCore* pCacheCore = new PointCacheCore();
	ctxt.PutUserData( (CValue::siPtrType)pCacheCore );
	
	return CStatus::OK;
}


SICALLBACK ReadPC2_Term( CRef& in_ref )
{
	Context ctxt = in_ref;
	CValue userData = ctxt.GetUserData( );
	if ( userData.IsEmpty() )
	{
		return CStatus::OK;
	}

	PointCacheCore* pCacheCore = (PointCacheCore*)(CValue::siPtrType)ctxt.GetUserData( );
	if ( pCacheCore )
	{
		delete pCacheCore;
	}

	ctxt.PutUserData( CValue() );
	
	return CStatus::OK;
}

SICALLBACK ReadPC2_BeginEvaluate( ICENodeContext& ctxt )
{
	ctxt.SetAsTimeVarying();

	CValue userData = ctxt.GetUserData( );
	if ( userData.IsEmpty() )
	{
		return CStatus::False;
	}

	PointCacheCore* pCacheCore = (PointCacheCore*)(CValue::siPtrType)ctxt.GetUserData( );
	if ( !pCacheCore )
	{
		return CStatus::False;
	}

	return CStatus::OK;

}


SICALLBACK ReadPC2_Evaluate( ICENodeContext& ctxt )
{


	ULONG out_portID = ctxt.GetEvaluatedOutputPortID( );

	CDataArrayString _strFilepath( ctxt, ID_IN_File_Name );
	CString strFilepath = _strFilepath[0];

	if(strFilepath.Length() == 0)
	{
		//Application().LogMessage(CString("No Cache File specified"), siWarningMsg);
		return CStatus::OK;
	}

	CDataArrayFloat _offset( ctxt, ID_IN_Offset );
	double offset = (double) _offset[0];

	CDataArrayFloat _timeWarp( ctxt, ID_IN_TimeWarp );
	double timeWarp = (double) _timeWarp[0];
	
	CDataArrayBool _customPlayback( ctxt, ID_IN_UseCustomPlayback );
	bool customPlayback = (bool) _customPlayback[0];

	CDataArrayFloat _customFrame( ctxt, ID_IN_Frame );
	double customFrame = (double) _customFrame[0];


	CDataArrayLong _interpolation( ctxt, ID_IN_Interpolation );
	int interpolation = (int) _interpolation[0];

	CDataArrayFloat _hermiteTension( ctxt, ID_IN_HermiteTension );
	double hermiteTension = (double) _hermiteTension[0];

	CDataArrayFloat _hermiteBias( ctxt, ID_IN_HermiteBias );
	double hermiteBias = (double) _hermiteBias[0];

	CDataArrayLong _loadOption( ctxt, ID_IN_LoadOption );
	int loadOption = (int) _loadOption[0];

	CDataArrayBool _useYUp( ctxt, ID_IN_UseYup );
	bool useYUp = (bool) _useYUp[0];

	
	int numDataPoints;
	double startFrame;
	double endFrame;
	double samples;
	int nbSamples;

	double currentFrame = ctxt.GetTime().GetTime();
	CVector3Array posArray(0);

	const wchar_t *orig = strFilepath.GetWideString();
	size_t origsize = wcslen(orig) + 1;
    const size_t newsize = 500;
    size_t convertedChars = 0;
    char nstring[newsize];
	wcstombs(nstring, orig, origsize);
	

	PointCacheCore* pCacheCore = (PointCacheCore*)(CValue::siPtrType)ctxt.GetUserData( );
	if ( pCacheCore && pCacheCore->OpenCache(nstring, loadOption) > -1 )
	{
		pCacheCore->GetNumPoints(numDataPoints);
		posArray.Resize((LONG)numDataPoints);

		pCacheCore->GetStartTime(startFrame);
		pCacheCore->GetSampleRate(samples);
		pCacheCore->GetNumSamples(nbSamples);

		endFrame = startFrame + (((double) nbSamples - 1) / samples);

		if( customPlayback )
		  currentFrame = customFrame;
		CalculatePosArray(pCacheCore, currentFrame, timeWarp, offset, posArray, useYUp, interpolation, hermiteTension, hermiteBias);
		if ( loadOption == 1 )
		  pCacheCore->Close();
	}
	else
	{
		//Application().LogMessage(CString("Unable to open Cache File <") + strFilepath + ">", siWarningMsg);
		return CStatus::OK;
	}

	switch( out_portID )
	{		
		case ID_OUT_Points :
		{

			if(posArray.GetCount() == 0)
			{
				return CStatus::OK;
			}

	
			CDataArray2DVector3f outData(ctxt);
			CIndexSet indexSet( ctxt );
			for(CIndexSet::Iterator it = indexSet.Begin(); it.HasNext(); it.Next())
			{
				CDataArray2DVector3f::Accessor outDataSubArray = outData.Resize(it, numDataPoints);
				for(int k = 0; k < numDataPoints; k++)
				{
					outDataSubArray[k] = CVector3f(posArray[k].GetX(), posArray[k].GetY(), posArray[k].GetZ());
				}
			
			}
		}
		break;

		case ID_OUT_StartFrame :
		{

			if(posArray.GetCount() == 0)
			{
				return CStatus::OK;
			}

	
			CDataArrayFloat outData(ctxt);
			outData[0] = (float) startFrame;
		}

		break;


		case ID_OUT_EndFrame :
		{

			if(posArray.GetCount() == 0)
			{
				return CStatus::OK;
			}

			CDataArrayFloat outData(ctxt);
			outData[0] = (float) endFrame;
		}

		break;

	};
	
	return CStatus::OK;
}

SICALLBACK ReadPC2_EndEvaluate( ICENodeContext& ctxt )
{

	CValue userData = ctxt.GetUserData( );
	if ( userData.IsEmpty() )
	{
		return CStatus::OK;
	}

	PointCacheCore* pCacheCore = (PointCacheCore*)(CValue::siPtrType)ctxt.GetUserData( );
	if ( pCacheCore )
	{
		pCacheCore->Close();
	}
	
	return CStatus::OK;
}
